stack/core/src/main/java/org/apache/usergrid/corepersistence/index/ReIndexServiceImpl.java

```diff
@@ -70,60 +79,228 @@
+
+        final Optional<ApplicationScope> appId = reIndexRequestBuilder.getApplicationScope();
 
-        final Observable<ApplicationScope>  applicationScopes = appId.isPresent()? Observable.just( getApplicationScope(appId.get()) ) : allApplicationsObservable.getData();
 
-        final String newCursor = StringUtils.sanitizeUUID( UUIDGenerator.newTimeUUID() );
+        Preconditions.checkArgument( !(cursor.isPresent() && appId.isPresent()),
+            "You cannot specify an app id and a cursor.  When resuming with cursor you must omit the appid" );
+
+        final Observable<ApplicationScope> applicationScopes = getApplications( cursor, appId );
+
+
+        final String jobId = StringUtils.sanitizeUUID( UUIDGenerator.newTimeUUID() );
+
+        final long modifiedSince = reIndexRequestBuilder.getUpdateTimestamp().or( Long.MIN_VALUE );
 
         //create an observable that loads each entity and indexes it, start it running with publish
-        final ConnectableObservable<EdgeScope> runningReIndex =
-            allEntityIdsObservable.getEdgesToEntities( applicationScopes, collectionName, startTimestamp )
+        final Observable<EdgeScope> runningReIndex = allEntityIdsObservable.getEdgesToEntities( applicationScopes,
+            reIndexRequestBuilder.getCollectionName(), cursorSeek.getSeekValue() )
+
+            //for each edge, create our scope and index on it
+            .doOnNext( edge -> {
+                final EntityIndexOperation entityIndexOperation = new EntityIndexOperation( edge.getApplicationScope(), edge.getEdge().getTargetNode(), modifiedSince );
+
+                logger.info( "Queueing {}", entityIndexOperation );
 
-                //for each edge, create our scope and index on it
-                .doOnNext( edge -> indexService.index( new EntityIdScope( edge.getApplicationScope(), edge.getEdge().getTargetNode() ) ) ).publish();
+                indexService.index(entityIndexOperation);
 
+            } );
 
 
         //start our sampler and state persistence
         //take a sample every sample interval to allow us to resume state with minimal loss
-        runningReIndex.sample( indexProcessorFig.getReIndexSampleInterval(), TimeUnit.MILLISECONDS,
-            rxTaskScheduler.getAsyncIOScheduler() )
-            .doOnNext( edge -> {
+        runningReIndex.buffer( indexProcessorFig.getUpdateInterval() )
+            //create our flushing collector and flush the edge scopes to it
+            .collect( () -> new FlushingCollector( jobId ),
+                ( ( flushingCollector, edgeScopes ) -> flushingCollector.flushBuffer( edgeScopes ) ) ).doOnNext( flushingCollector-> flushingCollector.complete() )
+                //subscribe on our I/O scheduler and run the task
+            .subscribeOn( Schedulers.io() ).subscribe();
+
+
+        return new ReIndexStatus( jobId, Status.STARTED, 0, 0 );
+    }
+
+
+    @Override
+    public ReIndexRequestBuilder getBuilder() {
+        return new ReIndexRequestBuilderImpl();
```

Comment:
Nah, Builder is just a bean, so we'll hide it's generation in the service, this way if it ultimately needs dependencies, we can wire them as required.